src/GPUActivityAgent.cpp

```diff
@@ -328,6 +342,26 @@ namespace geopm
             f_request = std::min(f_request, m_resolved_f_gpu_max);
             f_request = std::max(f_request, m_resolved_f_gpu_efficient);
 
+            if (phi >= 0.5) {
+                if (!std::isnan(gpu_utilization) &&
+                    gpu_utilization == 0) {
+                    if (m_gpu_idle_timer.at(domain_idx) > 0) {
+                        m_gpu_idle_timer.at(domain_idx) = m_gpu_idle_timer.at(domain_idx) - 1;
+                    }
+                }
+                else {
+                    // If no activity has been observed for a number of samples
+                    // IDLE_SAMPLE_COUNT we assume it is safe to reduce the frequency
+                    // to a minimum value.
+                    m_gpu_idle_timer.at(domain_idx) = IDLE_SAMPLE_COUNT;
+                }
+
+                if (m_gpu_idle_timer.at(domain_idx) <= 0) {
+                    f_request = m_freq_gpu_min;
```

Comment:
I think all of the code in this method that has anything to do with calculating f_request could be refactored into it's own helper method.  I started thinking about this because I'm seeing repeated code already:
```
                if (!std::isnan(gpu_utilization) &&
                    gpu_utilization == 0) {                 
```

I think this new code can be combined into the logic that starts on line 318 (new code), but further I think all of that can have an extract method refactor done on it to clean up ```adjust_platform()``` considerably, i.e. if the idle tracking time has expired and we're about to set the freq it to min, we don't need to do any of the other frequency calculation above.